lab_4_summarization_textrank/main.py

```diff
@@ -84,59 +113,99 @@ def __init__(self, stop_words: tuple[str, ...], punctuation: tuple[str, ...]) ->
+            raise ValueError
+        # чтобы вспомнить, что значит регулярка, см. lookahead и lookbehind
+        sentences = re.split(r'(?<=[?!.])\s+(?=[A-ZА-Я])', text)
+        tuple_with_sent = []
+        for count, value in enumerate(sentences):
+            sent = Sentence(value.replace('\n', ' ').replace('  ', ' '), count)
+            if sent:
+                tuple_with_sent.append(sent)
+        print(tuple_with_sent)
+        return tuple(tuple_with_sent)
+
 
     def _preprocess_sentences(self, sentences: tuple[Sentence, ...]) -> None:
         """
         Enriches the instances of sentences with their preprocessed versions
         :param sentences: a list of sentences
         :return:
         """
-        pass
+        if not check_type(sentences, tuple, Sentence):
+            raise ValueError
+        for sentence in sentences:
+            text = sentence.get_text()
+            new_sentence = super().preprocess_text(text)
+            sentence.set_preprocessed(new_sentence)
 
     def get_sentences(self, text: str) -> tuple[Sentence, ...]:
         """
         Extracts the sentences from the given text & preprocesses them
         :param text: the raw text
         :return:
         """
-        pass
-
+        if not isinstance(text, str):
+            raise ValueError
+        sentences = self._split_by_sentence(text)
+        self._preprocess_sentences(sentences)
+        return sentences
 
 class SentenceEncoder(TextEncoder):
     """
     A class to encode string sequence into matching integer sequence
     """
 
-    def __init__(self) -> None:
-        """
-        Constructs all the necessary attributes
-        """
-        pass
-
     def _learn_indices(self, tokens: tuple[str, ...]) -> None:
         """
         Fills attributes mapping words and integer equivalents to each other
         :param tokens: a sequence of string tokens
         :return:
         """
-        pass
+        if not check_type(tokens, tuple, str):
+            raise ValueError
+        if self._id2word:
+            start_v = max(self._id2word.keys())
+        else:
+            start_v = 1000
+        for count, token in enumerate(tokens, start=start_v):
+            if token not in (self._word2id.keys() and self._id2word.keys()):
+                self._word2id[token] = count
+                self._id2word[count] = token
 
     def encode_sentences(self, sentences: tuple[Sentence, ...]) -> None:
         """
         Enriches the instances of sentences with their encoded versions
         :param sentences: a sequence of sentences
         :return: a list of sentences with their preprocessed versions
         """
-        pass
+        if not check_type(sentences, tuple, Sentence):
+            raise ValueError
+        words = []
+        for sentence in sentences:
+            prepr = sentence.get_preprocessed()
+            for word in prepr:
+                words.append(word)
+        self._learn_indices(tuple(words))
```

Comment:
given the fact you can call `_learn_indices` more than once, consider calling it from each of the sentences

Reply:
done (?)

Reply:
not exaclty
I meant that you can just `_learn_indices(sentence.get_preprocessed())`

Reply:
done